contracts/0.4.24/nos/NodeOperatorsRegistry.sol

```diff
@@ -699,4 +787,158 @@ contract NodeOperatorsRegistry is INodeOperatorsRegistry, IsContract, AragonApp
+            operators[_nodeOperatorId].totalSigningKeys = usedSigningKeys; // discard unused keys
+            emit NodeOperatorTotalKeysTrimmed(_nodeOperatorId, trimmedKeys);
+        }
+    }
+
+    function _setTotalActiveKeys(uint64 _newTotalActiveKeys) internal {
+        keysUsageStats.totalActiveKeys = _newTotalActiveKeys;
+        emit ActiveKeysCountChanged(_newTotalActiveKeys);
+    }
+
+    function _updateTotalAvailableKeysCount() internal {
+        uint256 newTotalAvailableKeysCount = 0;
+        uint256 activeNodeOperatorsCount = getNodeOperatorsCount();
+        for (uint256 i = 0; i < activeNodeOperatorsCount; ++i) {
+            newTotalAvailableKeysCount += _getNodeOperatorAvailableKeysCount(i);
+        }
+        keysUsageStats.totalAvailableKeys = to64(newTotalAvailableKeysCount);
+        emit AvailableKeysCountChanged(newTotalAvailableKeysCount);
+    }
+
+    function _getNodeOperatorAvailableKeysCount(uint256 _operatorId) internal view returns (uint64) {
+        NodeOperator storage operator = operators[_operatorId];
+        if (!operator.active) {
+            return 0;
+        }
+        uint64 operatorStakingLimit = operator.stakingLimit;
+        uint64 operatorUsedSigningKeys = operator.usedSigningKeys;
+        uint64 operatorTotalSigningKeys = operator.totalSigningKeys;
+
+        // The keys limit of the validator is min(total signing keys, staking limit)
+        uint64 operatorSigningKeysLimit = operatorTotalSigningKeys > operatorStakingLimit ? operatorStakingLimit : operatorTotalSigningKeys;
+
+        return operatorUsedSigningKeys >= operatorSigningKeysLimit ? 0 : operatorSigningKeysLimit - operatorUsedSigningKeys;
+    }
+
+    function _setNodeOperatorStakingLimit(uint256 _id, uint64 _stakingLimit) internal {
+        require(operators[_id].stakingLimit != _stakingLimit, "NODE_OPERATOR_STAKING_LIMIT_IS_THE_SAME");
+        operators[_id].stakingLimit = _stakingLimit;
+        _updateTotalAvailableKeysCount();
+        emit NodeOperatorStakingLimitSet(_id, _stakingLimit);
+    }
+
+    /**
+     * @notice Return the initialized version of this contract starting from 0
+     */
+    function getVersion() external view returns (uint256) {
+        return CONTRACT_VERSION_POSITION.getStorageUint256();
+    }
+
+    /**
+     * @notice A function to finalize upgrade to v2 (from v1). Can be called only once
+     * @dev Value 1 in CONTRACT_VERSION_POSITION is skipped due to change in numbering
+     * For more details see https://github.com/lidofinance/lido-improvement-proposals/blob/develop/LIPS/lip-10.md
+     */
+    function finalizeUpgrade_v2(address _steth, bytes32 _type) external {
+        require(CONTRACT_VERSION_POSITION.getStorageUint256() == 0, "WRONG_BASE_VERSION");
+
+        _initialize_v2(_steth, _type);
+    }
+
+    function _initialize_v2(address _steth, bytes32 _type) internal {
+        require(_steth != address(0), "STETH_ADDRESS_ZERO");
```

Comment:
Move to `finalizeUpgrade_v2` to allow initialize with zero stETH for init from scratch.
Also worth to add setter for stETH